### ISSUE TITLE
Multi currency should not allow unsupported gateways to be selected.

### ISSUE THREAD

**Issue by dschaper** [impress-org/givewp#5655]
## User Story

I want payment gateways that do not support a currency to not display so that donors don't get to the final page of a multi-step form and see a huge warning that the currency is not supported or even worse they get a note with a donation value that is off by 1000%. They then click Back and get locked in to the default currency in the wrong amount.

## Details

Gateway currency support looks like it's only checked at the very last step, which is bad user experience. Some of the existing functionality breaks hard when a gateway is selected with an unsupported currency.

## Additional Context

PayPal has all currencies except for USD disabled. (No, enabling foreign currencies for PayPal is not a solution or an option, they hide exchange rates and mark up with outrageous fees.)

Select a currency that is not supported in the PayPal gateway:

![Screenshot_2021-02-26 Donate](https://user-images.githubusercontent.com/2782850/109332337-cd0f2b80-7812-11eb-8272-d1dc31b73408.png)

Fill in the donor information

![Screenshot_2021-02-26 Donate(1)](https://user-images.githubusercontent.com/2782850/109332336-cd0f2b80-7812-11eb-8d6e-b7ccd0a34bc7.png)

Select the PayPal gateway, get a fun error message.

![Screenshot_2021-02-26 Donate(2)](https://user-images.githubusercontent.com/2782850/109332335-cc769500-7812-11eb-9fed-1dcde2945026.png)

Go back to change the information, note the USD is now locked, but the buttons are still Euro.

![Screenshot_2021-02-26 Donate(3)](https://user-images.githubusercontent.com/2782850/109332334-cc769500-7812-11eb-9a93-ef8b910352f8.png)

Click a different donation button and get a completely different amount displayed.

![Screenshot_2021-02-26 Donate(4)](https://user-images.githubusercontent.com/2782850/109332329-cbddfe80-7812-11eb-986f-2c32891d0f9a.png)

## Acceptance Criteria

-   [x] When a currency is selected, disable all gateways that do not support that currency.
-   [x] Add error checking to make sure a value is NEVER an amount that is hundreds of percents more than a user entered.

**Comment by dschaper** [comment]
I don't think this is an enhancement. I think this is either a functionality that should have been part of the default, or it's a bug that needs to be addressed.

**Comment by kjohnson** [comment]
@dschaper that makes sense. Thanks for bringing this to our attention. I'll bring it to the table when we prioritize and delegate bug fixes at the end of the cycle.

**Comment by kjohnson** [comment]
Also, I've linked it to a new issue in the Currency Switcher private repository.

**Comment by dschaper** [comment]
Is the note on the plugin documentation page still accurate?

>Note: The GiveWP Currency Switcher add-on is not yet compatible with the new Multi-step form template in GiveWP 2.7. To use Currency Switcher, select legacy form templates. Learn more here.

**Comment by dschaper** [comment]
Any updates or information on this issue?

**Comment by JasonTheAdams** [comment]
@dschaper Thanks so much for providing all the great details for this. I've gone ahead and made a [Post on our feedback.givewp.com](https://feedback.givewp.com/bug-reports/p/multi-currency-should-not-allow-unsupported-gateways-to-be-selected) to track the request. This is the best way for us to prioritize these sorts of items and get them in future releases. This repository is best for tracking in progress work and developer questions (e.g. integrating with GiveWP). So I'm closing this Issue, but I don't want you to think we're just ignoring this. Please go over the the Post and upvote it. 😃

**Comment by dschaper** [comment]
Is there a way to see the repo for the plugin? If this is going to take another 3+ months to be considered or resolved then I'm happy to see if I can solve it myself.

**Comment by JasonTheAdams** [comment]
@dschaper I can appreciate how that would be confusing. For Bug Reports the votes don't reflect popularity, but is intended to give us a sense of how many customers are facing the issue. As customers report it to our support team, they upvote on behalf of customers. We use this to prioritize fixes.

I'm going to reach out to one of our developers tomorrow to get their thoughts on this fix. I feel as though it may be a bit more complex than it seems as it may not be so simple as a tweak to a single repo. Offhand, I'm not sure how we'd determine which gateways support the selected currency, as it really depends on the gateway and how it's configured. If it's just some work to be done in the Currency Switcher add-on, then I'll be happy to give you access to make a PR. In any case, thanks for offering!

**Comment by dschaper** [comment]
Thanks for the clarification.

I can manually set the Gateway <-> Currency Supported matrix. I would think there is some structure that stores that mapping and could be checked prior to selecting the Gateway options on Step 3 of the Multi-Step Form.
![Screenshot 2021-05-10 183759](https://user-images.githubusercontent.com/2782850/117745359-2d312f00-b1bf-11eb-940b-f0b69d8e4d78.png)

**Comment by JasonTheAdams** [comment]
Roger that, @dschaper!

@ravinderk Can you please glance over this and chat with me tomorrow?

**Comment by JasonTheAdams** [comment]
Hey @dschaper!

I spoke with Ravinder and came away with a few notes:
1. He does not recommend showing/hiding gateways based on supported gateways for a currency in multi-step, the reason being multi-step lives in an iframe and adjusts the height using some clever JS. Showing and hiding can have some strange side-effects when going to the final slide to select the gateway. That said, if you wanted to give it a shot and see what it does, I'd be fine with that, but you've been warned. 😆 
2. When a donor selects a non-supported gateway, an error message should already show up at the top of the form. This is how GiveWP currently handles this. Mind you, there is a crazy styling issue on multi-step I just found.
3. There are other problems with the whole flow which are, frankly, even worse than what we've discussed here — as it's _very_ difficult for a donor to select a supported gateway once they've selected an unsupported one.

I put together a video (below) to share with my team so we have a full view of all the issues going on. We work in 6-week cycles, one of which is starting in two weeks. After each cycle we have 2 weeks for miscellaneous work, which this would be a good one for. Given the strange flow issues, this will likely take a bit of planning, too. It's slightly possible we could knock this out next week, but I wouldn't bet on it. It's more likely this will take ~8 weeks to get to. If you'd like to take a stab at it yourself before then, I'm happy to give you access to the Currency Switcher repository to make a PR (it may take a PR in both GiveWP core and Currency Switcher).

https://user-images.githubusercontent.com/2024145/117869343-88126700-b24f-11eb-9761-9feaf96184fb.mp4

**Comment by Benunc** [comment]
Hey @dschaper thanks for reaching back out. The reason this GitHub issue was closed is that we use our feedback site for issues that are not in active development, and still need for the product management team to come up with a plan. We created a feedback post (which I see you've commented on) that has all of your (very helpful!) insights as an internal note for our team to have a look. I am going to make sure that this is at least discussed as a part of the next cycle, but we have lots of feature requests, enhancements, and other development needs that have to be balanced as a part of the decision. 

GitHub issues represent _what the team is currently actively working on_ and so to leave this as an open issue is not helpful. The feedback site is for issues that have not been prioritized yet. Stay tuned there for the latest. 

Also, this particular issue (once it's prioritized and created) will more accurately live in the Currency Switcher repository, which is a private repository here, not publicly visible. 

Thanks again for the very helpful insights into the problem.
